### PR TITLE
Glooctl license check

### PR DESCRIPTION
**Pull request by asayah**
# Description

Check the validity of Gloo Gateway License using `glooctl check license --license-key <key>`.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: docs/content/reference/cli/glooctl_check_license.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  -i, --interactive                       use interactive mode
+      --kube-context string               kube context to use when interacting with kubernetes
+      --kubeconfig string                 kubeconfig to use, if not standard one
+  -n, --namespace string                  namespace for reading or writing resources (default "gloo-system")
+  -o, --output OutputType                 output format: (json, table) (default table)
+  -p, --pod-selector string               Label selector for pod scanning (default "gloo")
+      --read-only                         only do checks that dont require creating resources (i.e. port forwards)
+  -r, --resource-namespaces stringArray   Namespaces in which to scan gloo custom resources. If not provided, all watched namespaces (as specified in settings) will be scanned.
+      --use-consul                        use Consul Key-Value storage as the backend for reading and writing config (VirtualServices, Upstreams, and Proxies)
+```
```
Comment (sam-heilbron) [inline docs/content/reference/cli/glooctl_check_license.md:46]: A challenge/side-effect with making the license check a sub-command of the check command is that all of these parameters which dont have an effect on the license, will be available to users. Would we ever want to have a `glooctl license validate` or something along those lines?

**Review by sam-heilbron — COMMENTED** [review]
Just one question about how/where we want to structure this command.

Thanks for the contribution!

File: docs/content/reference/cli/glooctl_check_license.md (hunk 1 above)
Comment (asayah) [reply docs/content/reference/cli/glooctl_check_license.md:46]: yes I think we can have 
a `glooctl license validate` 
maybe a `glooctl license update` in the future ....

Comment (asayah) [reply docs/content/reference/cli/glooctl_check_license.md:46]: will refactor

Comment (asayah) [reply docs/content/reference/cli/glooctl_check_license.md]: @sam-heilbron  made the change now it is
`license validate --license-key`

though I don't see an effective way of removing the global flags coming from the top root (unless I move it to the top top https://github.com/solo-io/gloo/blob/main/projects/gloo/cli/pkg/cmd/root.go#L72)

 wdyt? fine to keep it this way or should I make the change?

Comment (sam-heilbron) [reply docs/content/reference/cli/glooctl_check_license.md]: I think it's good as is! This will be a useful enhancement

**Review by sam-heilbron — APPROVED** [review]

File: projects/gloo/cli/pkg/cmd/license/root.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+func RootCmd(opts *options.Options, optionsFunc ...cliutils.OptionsFunc) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "license",
+		Aliases: []string{"l"},
+		Short:   "subcommands for interacting with the license",
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			if err := prerun.CallParentPrerun(cmd, args); err != nil {
+				return err
+			}
+			if err := prerun.EnableConsulClients(opts); err != nil {
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/license/root.go:19]: why is this needed?

File: projects/gloo/cli/pkg/cmd/license/route_suite_test.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package license
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+func TestRoute(t *testing.T) {
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/license/route_suite_test.go:10]: should this file be renamed license_suite_test?

and all the Route here should say License instead

File: projects/gloo/cli/pkg/cmd/license/validate_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	It("should fail to verify an invalid license", func() {
+		invalidLicenseKey := "invalid"
+
+		// Call the checkLicense function
+		err := checkLicense(invalidLicenseKey)
+		Expect(err).To(HaveOccurred())
+		Expect(err.Error()).To(Equal("can't parse licence key"))
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/license/validate_test.go:43]:
```suggestion
		Expect(err.Error()).To(Equal("can't parse license key"))
```
nit: consistent spelling

File: projects/gloo/cli/pkg/cmd/license/validate.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	cmd.MarkFlagRequired(flagutils.LicenseFlag)
+	return cmd
+}
+
+func checkLicense(licenseKey string) error {
+	var licenseClaims LicenseClaims
+
+	_, _, err := new(jwt.Parser).ParseUnverified(licenseKey, &licenseClaims)
+	if err != nil {
+		return fmt.Errorf("can't parse licence key")
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/license/validate.go:68]:
```suggestion
		return fmt.Errorf("can't parse license key")
```

File: projects/gloo/cli/pkg/cmd/license/validate.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	fmt.Printf(formatLicenseDetail(licenseClaims.CreationDate, licenseClaims.ExpirationDate, licenseClaims.Product, licenseClaims.LicenseType == "trial"))
+	return nil
+}
+
+func checkLegacyLicense(licenseKey string) error {
+	var licenseLegacyClaim LicenseLegacyClaims
+	var standardizedLicenseKey = base64.RawURLEncoding.EncodeToString([]byte(`{"alg":"HS256","typ":"JWT"}`)) + "." + licenseKey
+	_, _, err := new(jwt.Parser).ParseUnverified(standardizedLicenseKey, &licenseLegacyClaim)
+	if err != nil {
+		return fmt.Errorf("can't parse licence key")
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/license/validate.go:79]:
```suggestion
		return fmt.Errorf("can't parse license key")
```

File: projects/gloo/cli/pkg/cmd/license/validate_test.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	"time"
+
+	"github.com/golang-jwt/jwt/v4"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Licence", func() {
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/license/validate_test.go:11]:
```suggestion
var _ = Describe("License Validate", func() {
```

File: projects/gloo/cli/pkg/cmd/options/options.go [hunk 8]
```diff
@@ -18,23 +18,24 @@ import (
...
+	Edit         Edit
+	Route        Route
+	Get          Get
+	Add          Add
+	Istio        Istio
+	Remove       Remove
+	Cluster      Cluster
+	Check        Check
+	CheckCRD     CheckCRD
+	CheckLicense CheckLicense
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/options/options.go:38]: should this be renamed License (or LicenseValidate) to match the command?

File: projects/gloo/cli/pkg/cmd/license/root.go (hunk 2 above)
Comment (asayah) [reply projects/gloo/cli/pkg/cmd/license/root.go:19]: good catch, not needed, used "route" as a base

**Review by jenshu — APPROVED** [review]
